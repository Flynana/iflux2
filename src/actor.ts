/**
 * Actor，致敬Erlang，Scala的akka的Actor model
 * Actor, 独立计算的执行单元
 * 我们不共享状态(share state), 只去transform state
 */
import { OrderedMap } from 'immutable'

type IMap = OrderedMap<string, any>;
type Route = {
  [name: string]: (state: IMap, params?: any) => IMap
};

export default class Actor {
  //记录当前的路由信息
  _route: Route;

  /**
   * 定义actor的默认状态
   * @returns {{}}
   */
  defaultState(): Object {
    return {};
  }

  /**
   * actor的mode的receive,被store在dispatch的时候调用
   * @param msg
   * @param state
   * @param param
   * @returns {Object}
   */
  receive(msg: string, state: IMap, param?: any): IMap {
    //this._route是在@Action标记中初始化完成
    const route = this._route || {};
    //获取处理的函数
    const fn = route[msg];

    //如果可以处理返回处理后的结果，否则直接返回state
    return fn ? fn.call(this, state, param) : state;
  }
}
