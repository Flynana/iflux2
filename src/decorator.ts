/**
 * Action decorator
 *
 * 用于标记Actor中的实力方法，主要的作用是给Actor绑定当前的handler方法
 * 便于Actor的receive方法可以搜索到哪个handler可以处理dispatch过来的事件
 *
 * Usage:
 * import {Actor, Action} from 'iflux2'
 *
 * class HelloActor extends Actor {
 *   defaultState() {
 *    return {text: 'hello iflxu2'}
 *   }
 *
 *   @Action('change')
 *   change(state, text) {
 *    return state.set('text', text);
 *   }
 *
 * }
 *
 *
 * @param msg 事件名称
 * @constructor
 */
export const Action = (msg: string) => (
  target: any,
  props: any,
  descriptor: TypedPropertyDescriptor<any>
) => {
  target._route = target._route || {};
  target._route[msg] = descriptor.value;
};


/**
 * 动态的绑定组件的上下文
 *
 * Usage
 * import React, {Component} from 'iflux2'
 *
 * @CtxStoreName('_store')
 * class Hello extends Component {
 *   render() {
 *    return <div>hello world</div>
 *   }
 * }
 * @param obj 绑定上下文
 * @returns {function(Object)}
 */
export const CtxStoreName = (name: string) => {
  return (target: Object) => {
    //动态的在组件的上下午绑定storeName
    //通过该标记反向的告诉Relax获取正取的上下文store
    target['_ctxStoreName'] = name;
  };
};
