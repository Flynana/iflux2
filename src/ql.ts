/**
 * 查询语言
 * @flow
 */

'use strict';

import { isQuery } from './util';

type Lang = Array<any>;

//递增的id
let uuid = 0;

export class QueryLang {
  _id: number;
  _lang: Lang;
  _name: string;

  /**
   * init
   */
  constructor(name: string, lang: Lang) {
    this._id = ++uuid;
    this._name = name;
    this._lang = lang;

    if (process.env.NODE_ENV != 'production') {
      if (!isQuery(this._lang)) {
        throw new Error(`${this._name} is not invalid`)
      }
    }
  }

  /**
   * 判断当前是不是一个合法的query lang
   * @returns {boolean}
   */
  isValidQuery(): boolean {
    return isQuery(this._lang);
  }

  /**
   * 当前的id
   * @returns {number}
   */
  id(): number {
    return this._id;
  }

  /**
   * 当前的name
   */
  name(): string {
    return this._name;
  }

  /**
   * 当前的语法标记
   * @returns {Array.<Object>}
   */
  lang() {
    return this._lang;
  }

  setLang(lang: Lang) {
    this._lang = lang;
    return this;
  }
}

//export factory method
export const QL = (
  name: string,
  lang: Lang
) => new QueryLang(name, lang);
