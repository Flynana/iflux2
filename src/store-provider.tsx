/**
 * StoreProvider
 * 主要的作用是在Store和React的App之间建立桥梁
 * 将Store初始化,切绑定到React顶层App的上下文
 */

import * as React from 'react';
import Store from './store';

type TStore = typeof Store;
type StoreOptions = {
  debug?: boolean;
  ctxStoreName?: string;
};

/**
 * WrapperComponent
 * @param AppStore
 * @param opts
 * @returns {Function}
 */
export default function connectToStore(
  AppStore: TStore,
  opts: StoreOptions = { debug: false, ctxStoreName: '_iflux2$store' }
): React.Component {
  return function (Component: React.Component) {
    //获取上下午动态设置的store的名称
    //避免Relax在获取context的时候就近原则的冲突
    const ctxStoreName = opts.ctxStoreName || '_iflux2$store';
    //proxy Component componentDidMount
    const proxyDidMount = Component.prototype.componentDidMount || (() => { });
    //清空
    Component.prototype.componentDidMount = () => { };

    return class StoreContainer extends React.Component {
      props: Object;
      //关联的store
      _store: Store;
      //当前的组件状态
      _isMounted: boolean;
      //获取当前的ref
      App: Object;

      static displayName = `StoreProvider(${getDisplayName(Component)})`;

      static childContextTypes = {
        [ctxStoreName]: React.PropTypes.object
      };

      getChildContext: Function = (): Object => {
        return {
          [ctxStoreName]: this._store
        };
      };

      constructor(props: Object) {
        super(props);
        if (process.env.NODE_ENV != 'production') {
          //如果是debug状态
          if (opts.debug) {
            console.group && console.group(`StoreProvider(${Component.name}) in debug mode.`);
            console.time('first-render-time');
          }
        }

        //初始化当前的组件状态
        this._isMounted = false;
        //初始化Store
        this._store = new AppStore(opts);
        this._store.subscribe(this._handleStoreChange);
      }

      componentDidMount() {
        if (process.env.NODE_ENV != 'production') {
          if (opts.debug) {
            console.timeEnd('first-render-time');
            console.groupEnd && console.groupEnd();
          }
        }

        this._isMounted = true;

        //代理的子componentDidMount执行一次
        if (this.App) {
          proxyDidMount.call(this.App);
        }
      }

      componentWillUpdate() {
        this._isMounted = false;

        if (process.env.NODE_ENV != 'production') {
          if (opts.debug) {
            console.group && console.group(`StoreProvider(${Component.name}) will update`);
            console.time('update-render-time');
          }
        }
      }

      componentDidUpdate() {
        this._isMounted = true;

        if (process.env.NODE_ENV != 'production') {
          if (opts.debug) {
            console.timeEnd('update-render-time');
            console.groupEnd && console.groupEnd();
          }
        }
      }

      componentWillUnmount() {
        this._store.unsubscribe(this._handleStoreChange);
      }

      render() {
        return (
          <Component
            ref={(App) => this.App = App}
            {...this.props}
            store={this._store}
          />
        );
      }


      _handleStoreChange = () => {
        if (this._isMounted) {
          (this as any).forceUpdate();
        }
      };
    }
  };

  function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component'
  }
}
