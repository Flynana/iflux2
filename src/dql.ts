
import { QueryLang } from './ql';
import { isArray, isStr, isQuery } from './util';

type Lang = Array<any>;

/**
 * 动态的QueryLang
 */
export class DynamicQueryLang {
  _ctx: Object;
  _name: string;
  _lang: Lang;

  constructor(name: string, lang: Lang) {
    this._ctx = {};
    this._name = name;
    this._lang = lang;

    if (process.env.NODE_ENV != 'production') {
      if (!isQuery(this._lang)) {
        throw new Error(`${this._name} syntax error`)
      }
    }
  }

  /**
   * 分析路径中的动态元素，然后根据上下文替换
   * @param ql
   */
  analyserLang(dLang: Lang) {
    const lang = []

    for (let i = 0, len = dLang.length; i < len; i++) {
      //获取当前的路径
      let path = dLang[i];

      if (isStr(path) && path[0] === '$') {
        lang[i] = path[0] === '$' ? this._ctx[path.substring(1)] : path;
      } else if (isArray(path)) {
        //init
        lang[i] = []
        for (let j = 0, len = path.length; j < len; j++) {
          let field = dLang[i][j];
          lang[i][j] = isStr(field) && field[0] === '$' ? this._ctx[field.substring(1)] : field;
        }
      } else if (path instanceof DynamicQueryLang) {
        lang[i] = new QueryLang(path._name + '2QL', this.analyserLang(path._lang));
      } else {
        //zero runtime cost
        if (process.env.NODE_ENV != 'production') {
          //如果path是QueryLang，校验querylang语法的合法性
          if (path instanceof QueryLang && !path.isValidQuery()) {
            throw new Error(`DQL: syntax error`);
          }
        }
        lang[i] = path;
      }
    }


    return lang;
  }

  /**
   * 设置上下文
   * @param  {Object} ctx
   */
  context(ctx: Object) {
    this._ctx = ctx;
    return this;
  }

  name() {
    return this._name
  }

  lang() {
    return this._lang
  }
}

/**
 * 工厂函数
 * @param  {string name
 * @param  {Lang} lang
 */
export const DQL = (
  name: string,
  lang: Lang
) => new DynamicQueryLang(name, lang);
