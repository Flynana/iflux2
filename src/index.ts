import msg from './msg';
import Actor from './actor';
import StoreProvider from './store-provider';
import Relax from './relax';
import Store from './store';
import { Action, CtxStoreName } from './decorator';
import { QL } from './ql';
import { DQL } from './dql';

export {
  QL,
  DQL,
  msg,
  Action,
  CtxStoreName,
  Actor,
  Relax,
  Store,
  StoreProvider
};
