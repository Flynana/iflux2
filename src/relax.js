/**
 * 致敬Reley,更希望我们小伙伴可以relax
 *
 * Relax根据containerComponent的defaultProps
 * 自动数据依赖注入, 数据源优先级为:
 * 1. this.props
 * 2. store的action函数
 * 3. query-lang
 * 4. store的state
 * 5. 组件设置的默认值
 *
 * @flow
 */

'use strict';

import React from 'react';
import { fromJS } from 'immutable';
import assign from 'object-assign';

import { QueryLang } from './ql';
import { DynamicQueryLang } from './dql';

import type {IState } from './types'

type State = {
  storeState: IState;
};

type Store = {
  _debug: boolean;
  state(): IState;
  bigQuery(ql: mixed): any;
  subscribe: Function;
  unsubscribe: Function;
};

export default function Relax(
  Component: ReactClass<{}>
): ReactClass<{}> {
  //获取组件中绑定的上下文storeName的参数
  //默认是store
  const ctxStoreName = Component._ctxStoreName || '_iflux2$store';
  return class RelaxContainer extends React.Component {
    //当前的状态
    state: State;
    //当前组件的挂载状态
    _isMounted: boolean;
    //当前的所有的子组件的props
    _relaxProps: Object;
    //debug状态
    _debug: boolean;
    //当前上下文的store
    _store: Store;
    _dql2ql: Object;

    //声明上下文类型
    static contextTypes = {
      [ctxStoreName]: React.PropTypes.object
    };

    //声明displayName
    static displayName = `Relax(${getDisplayName(Component)})`;

    constructor(props) {
      super(props);
      //当前组件的挂载状态
      this._isMounted = false;
      this._dql2ql = {};
      //当前组件的状态
      this.state = {
        storeState: fromJS({})
      };
    }

    componentWillMount() {
      //设置当前组件的状态
      this._isMounted = false;
      this._store = this.context[ctxStoreName];

      //检查store是不是存在上下文
      //抛出异常方便定位问题
      if (!this._store) {
        throw new Error('Could not find any @StoreProvider bind AppStore in current context');
      }

      //在开发阶段可以有更好的日志跟踪，在线上可以drop掉log，reduce打包的体积
      if (process.env.NODE_ENV != 'production') {
        if (this._store._debug) {
          console.time('relax time');
          console.groupCollapsed(`Relax(${Component.name}) will mount`);
        }
      }

      //计算最终的props,这样写的是避免querylang的重复计算
      this._relaxProps = assign({}, this.props, this.getProps(this.props));

      if (process.env.NODE_ENV != 'production') {
        //trace log
        if (this._store._debug) {
          console.timeEnd('relax time');
          console.groupEnd();
        }
      }
    }

    componentDidMount() {
      this._isMounted = true;
      this._store.subscribe(this._subscribeStoreChange);
    }

    componentWillUpdate() {
      this._isMounted = false;
    }

    componentDidUpdate() {
      this._isMounted = true;
    }

    componentWillUnmount() {
      this._store.unsubscribe(this._subscribeStoreChange);
    }

    /**
     * 3ks immutable
     * @param nextProps
     * @returns {boolean}
     */
    shouldComponentUpdate(nextProps: Object) {
      //will drop
      if (process.env.NODE_ENV != 'production') {
        if (this._store._debug) {
          console.time('relax time');
          console.groupCollapsed(`Relax(${Component.name}) should update`);
        }
      }

      //compare nextProps && this.props
      //如果属性不一致, 就去render
      if (Object.keys(nextProps).length != Object.keys(this.props).length) {
        if (process.env.NODE_ENV != 'production') {
          if (this._store._debug) {
            console.groupEnd();
          }
        }

        return true;
      }

      //合并新的属性集合
      //判断是不是数据没有变化, 如果没有变化不需要render
      const newRelaxProps = assign({}, nextProps, this.getProps(nextProps));

      for (let key in newRelaxProps) {
        if (newRelaxProps.hasOwnProperty(key)) {
          if (newRelaxProps[key] != this._relaxProps[key]) {
            this._relaxProps = newRelaxProps;

            if (process.env.NODE_ENV != 'production') {
              //trace log
              if (this._store._debug) {
                console.timeEnd('relax time');
                console.groupEnd();
              }
            }

            return true;
          }
        }
      }

      if (process.env.NODE_ENV != 'production') {
        if (this._store._debug) {
          console.log(`Relax(${Component.name}) avoid re-render`);
          console.timeEnd('relax time');
          console.groupEnd();
        }
      }

      return false;
    }

    render() {
      return (
        <Component {... this._relaxProps} />
      );
    }

    /**
     * 计算prop的值 然后自动注入
     *
     * 1. 默认属性是不是存在，不存在返回空对象
     * 2. 默认属性的值是不是一个合法的query-lang， 如果是就在store中通过bigQuery计算
     * 3. 默认属性是不是在父组件传递的props中，如果是取
     * 4. 是不是store得属性
     * 5. 是不是store得某个key值
     * 6. 都不是就是默认值
     */
    getProps(reactProps) {
      const dql = {};
      const props = {};
      const store = this._store;
      const defaultProps = Component.defaultProps || {};

      for (let propName in defaultProps) {
        if (defaultProps.hasOwnProperty(propName)) {

          //判断defaultProps的值是不是query的语法
          const propValue = defaultProps[propName];
          if (propValue instanceof QueryLang) {
            props[propName] = store.bigQuery(propValue);
            continue;
          }

          //隔离出来DQL
          if (propValue instanceof DynamicQueryLang) {
            dql[propName] = propValue;

            //如果不存在转换，创建一个QL与关联
            if (!this._dql2ql[propName]) {
              //这个lang实际上并不是QueryLang需要的
              //这个lang会被后面真正被DynamicQueryLang计算过的lang取代
              this._dql2ql[propName] = new QueryLang(propValue.name(), propValue.lang());
            }

            continue;
          }

          props[propName] = defaultProps[propName];

          //如果默认属性中匹配上
          if (RelaxContainer._isNotUndefinedAndNull(reactProps[propName])) {
            props[propName] = reactProps[propName];
          } else if (RelaxContainer._isNotUndefinedAndNull(store[propName])) {
            props[propName] = store[propName];
          } else if (RelaxContainer._isNotUndefinedAndNull(store.state().get(propName))) {
            props[propName] = store.state().get(propName);
          }
        }
      }

      //开始计算DQL
      for (let propName in dql) {
        if (dql.hasOwnProperty(propName)) {
          //取出dynamicQL
          const dqlObj = dql[propName];
          const lang = dqlObj.context(props).analyserLang(dqlObj.lang())
          const ql = this._dql2ql[propName].setLang(lang)
          props[propName] = store.bigQuery(ql);
        }
      }

      return props;
    }

    /**
     * 判断当前的值是不是undefined或者null
     * @param  {any} param
     */
    static _isNotUndefinedAndNull(param: any) {
      return typeof (param) != 'undefined' && null != param;
    }

    /**
     * 订阅store的变化
     */
    _subscribeStoreChange = (state: IState) => {
      if (this._isMounted) {
        //re-render
        this.setState({ storeState: state });
      }
    };
  };

  /**
   * displayName
   */
  function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component'
  }
}
