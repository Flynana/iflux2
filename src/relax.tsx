/**
 * 致敬Reley,更希望我们小伙伴可以relax
 *
 * Relax根据containerComponent的defaultProps
 * 自动数据依赖注入, 数据源优先级为:
 * 1. this.props
 * 2. store的action函数
 * 3. query-lang
 * 4. store的state
 * 5. 组件设置的默认值
 */
import * as React from 'react';
import { is, fromJS, OrderedMap } from 'immutable';
import Store from './store'
import { QueryLang } from './ql';
import { DynamicQueryLang } from './dql';

type State = OrderedMap<string, any>;
type RelaxContext = {
  [name: string]: Store;
};

export default function Relax(Component: React.Component): React.Component {
  //获取组件中绑定的上下文storeName的参数
  //默认是store
  const ctxStoreName = Component._ctxStoreName || '_iflux2$store';

  return class RelaxContainer extends React.Component {
    //当前的状态
    state: Object;
    //当前的属性
    props: Object;
    //当前上下文的类型
    context: RelaxContext;

    //debug状态
    _debug: boolean
    //当前组件的挂载状态
    _isMounted: boolean;;
    //当前上下文的store
    _store: Store;
    //缓存当前的dql2ql
    _dql2ql: Object;
    //当前的所有的子组件的props
    _relaxProps: Object;

    //声明上下文类型
    static contextTypes = {
      [ctxStoreName]: React.PropTypes.object
    };

    //声明displayName
    static displayName = `Relax(${getDisplayName(Component)})`;

    constructor(props: Object, context: RelaxContext) {
      super(props);

      this._dql2ql = {};
      this._isMounted = false;
      this.state = { storeState: fromJS({}) }
      this._store = context[ctxStoreName];
      this._debug = this._store._debug;

      this._store.subscribe(this._subscribeStoreChange);
    }

    componentWillMount() {
      //设置当前组件的状态
      this._isMounted = false;

      //检查store是不是存在上下文
      //抛出异常方便定位问题
      if (!this._store) {
        throw new Error('Could not find any @StoreProvider bind AppStore in current context');
      }

      //计算最终的props,这样写的是避免querylang的重复计算
      this._relaxProps = this.computedRelaxProps(this.props);

      //在开发阶段可以有更好的日志跟踪，在线上可以drop掉log，reduce打包的体积
      if (process.env.NODE_ENV != 'production') {
        if (this._debug) {
          console.groupCollapsed && console.groupCollapsed(`Relax(${Component.name}) will mount 🚀`);
          console.log('props:|>', JSON.stringify(this.props, null, 2));
          console.log('relaxProps:|>', JSON.stringify(this._relaxProps, null, 2));
          console.groupEnd && console.groupEnd();
        }
      }
    }

    componentDidMount() {
      this._isMounted = true;
    }

    componentWillUpdate() {
      this._isMounted = false;
    }

    componentDidUpdate() {
      this._isMounted = true;
    }

    componentWillUnmount() {
      this._store.unsubscribe(this._subscribeStoreChange);
    }

    /**
     * 3ks immutable
     * @param nextProps
     * @returns {boolean}
     */
    shouldComponentUpdate(nextProps: Object) {
      const newRelaxProps = this.computedRelaxProps(nextProps);

      if (
        !is(fromJS(newRelaxProps), fromJS(this._relaxProps)) ||
        !is(fromJS(this.props), fromJS(nextProps))) {
        this._relaxProps = newRelaxProps;

        //log trace        
        if (process.env.NODE_ENV != 'production') {
          if (this._debug) {
            console.groupCollapsed(`Relax(${Component.name}) will update 🚀`);
            console.log('props:|>', JSON.stringify(nextProps, null, 2));
            console.log('relaxProps:|>', JSON.stringify(this._relaxProps, null, 2));
            console.groupEnd();
          }
        }

        return true;
      }

      return false;
    }

    render() {
      return (
        <Component {...this.props} {... this._relaxProps} />
      );
    }

    /**
     * 计算prop的值 然后自动注入
     *
     * 1. 默认属性是不是存在，不存在返回空对象
     * 2. 默认属性的值是不是一个合法的query-lang， 如果是就在store中通过bigQuery计算
     * 3. 默认属性是不是在父组件传递的props中，如果是取
     * 4. 是不是store得属性
     * 5. 是不是store得某个key值
     * 6. 都不是就是默认值
     */
    computedRelaxProps(reactProps) {
      const dql = {} as { [name: string]: DynamicQueryLang };
      const relaxProps = {};
      const store = this._store;
      const defaultProps = Component.defaultProps || {};

      for (let propName in defaultProps) {
        //获取当前的属性值
        const propValue = defaultProps[propName];
        //先默认值
        relaxProps[propName] = propValue;

        //判断defaultProps的值是不是query的语法
        if (propValue instanceof QueryLang) {
          relaxProps[propName] = store.bigQuery(propValue);
          continue;
        }

        //隔离出来DQL
        else if (propValue instanceof DynamicQueryLang) {
          dql[propName] = propValue;

          //如果不存在转换，创建一个QL与关联
          if (!this._dql2ql[propName]) {
            //这个lang实际上并不是QueryLang需要的
            //这个lang会被后面真正被DynamicQueryLang计算过的lang取代
            this._dql2ql[propName] = new QueryLang(propValue.name(), propValue.lang());
          }

          continue;
        }

        //如果默认属性中匹配上
        if (RelaxContainer._isNotValidValue(reactProps[propName])) {
          relaxProps[propName] = reactProps[propName];
        } else if (RelaxContainer._isNotValidValue(store[propName])) {
          relaxProps[propName] = store[propName];
        } else if (RelaxContainer._isNotValidValue(store.state().get(propName))) {
          relaxProps[propName] = store.state().get(propName);
        }
      }

      //开始计算DQL
      for (let propName in dql) {
        if (dql.hasOwnProperty(propName)) {
          //取出dynamicQL
          const dqlObj = dql[propName];
          const lang = dqlObj.context(relaxProps).analyserLang(dqlObj.lang())
          const ql = this._dql2ql[propName].setLang(lang)
          relaxProps[propName] = store.bigQuery(ql);
        }
      }

      return relaxProps;
    }

    /**
     * 判断当前的值是不是undefined或者null
     * @param  {any} param
     */
    static _isNotValidValue(param: any) {
      return typeof (param) != 'undefined' && null != param;
    }

    /**
     * 订阅store的变化
     */
    _subscribeStoreChange = (state: State) => {
      if (this._isMounted) {
        (this as any).setState({
          storeState: state
        })
      }
    };
  };

  /**
   * displayName
   */
  function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component'
  }
}
