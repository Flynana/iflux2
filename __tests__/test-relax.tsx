import * as React from 'react'
import * as renderer from 'react-test-renderer';
import { Actor, Action, Store, Relax, StoreProvider } from "../src/index";
jest.mock('react-dom');

//;;;;;;;;;;Actor;;;;;;;;;;;;;;;;;;
class TestActor extends Actor {
  defaultState() {
    return {
      text: 'hello world'
    }
  }

  @Action('init')
  init(state, text) {
    return state.set('text', text);
  }
}


//;;;;;;;;;;;;;;;Store;;;;;;;;;;;;;;;
class AppStore extends Store {
  bindActor() {
    return [
      new TestActor()
    ]
  }

  init = () => {
    this.dispatch('init', 'hello iflux2');
  };
}

//;;;;;;;;;;;;;;;;Relax;;;;;;;;;;;;;;;
@Relax
class Text extends React.Component {
  props: {
    text: string;
  };

  static defaultProps = {
    text: ''
  };

  render() {
    expect(this.props.text).toEqual('hello iflux2');
    return (
      <div>{this.props.text}</div>
    )
  }
}

//;;;;;;;;;;;;;;;;;;root;;;;;;;;;;;;;
@StoreProvider(AppStore)
class HelloApp extends React.Component {
  props: {
    store: AppStore;
  };

  componentWillMount() {
    this.props.store.init();
  }

  render() {
    return (
      <Text />
    )
  }
}

//;;;;;;;;;;;test;;;;;;;;;;;;;;;;;;;;;
test('test store provider and relax', () => {
  const tree = renderer.create(<HelloApp />).toJSON();
  expect(tree).toMatchSnapshot();
});