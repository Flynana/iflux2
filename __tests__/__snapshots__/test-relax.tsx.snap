// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test store provider and relax 1`] = `
<div>
  hello iflux2
</div>
`;
